### ISSUE TITLE
The profile cannot verify a signature on the message

### ISSUE THREAD

**Issue by thiema** [latchset/mod_auth_mellon#30]
Hello

I've the follwowing issue while running mod_auth_mellon against an Enterprise ADFS. I understand this here isnt a support forum, but I had no luck to login and subscribing to https://sympa.uninett.no/lists/uninett.no/subscribe/modmellon. The error I enconter while trying to subsribe is " ERROR (subscribe) - Authorization rejected. Maybe you forgot to log in? ". However, I can say, that I am already registered and logged in.

Back to my issue with mod_auth_mellon:
I already configured SP and (together with my colleage responsible for ADFS configuration) IdP. Now, when testing the SSO I can see following error:

[APLOG_ERR auth_mellon_handler.c:2203] Error processing authn response. Lasso error: [440] The profile cannot verify a signature on the message, SAML Response: StatusCode1="urn:oasis:names:tc:SAML:2.0:status:Success", StatusCode2="(null)", StatusMessage="(null)"

According to this guide https://jdennis.fedorapeople.org/doc/mellon-user-guide/mellon_user_guide.html#web_sso_flow I am failing at step 4 (IdP responds with <assertion> embedded in form...)

Before presenting more details, please guide me:
1. Where is correct place to post that issue ?
2. How to properly subscribe: (-> https://sympa.uninett.no/lists/uninett.no/subscribe/modmellon)
3. Which details to provide in order to further analyse and understand the isssue.

Regards
Marian

Edit1:
Versions:
mod_auth_mellon: latest (master branch)
lasso: 2.5.99
httpd: 2.4.34

**Comment by thijskh** [comment]
1. & 2.: I've asked the list owner about the problems with the mailing list.

3.: If you get a signature failure I'd first investigate whether you're configured the correct assertion signing certificate from the IdP in mellon. Perhaps it helps to enhance logging options / diagnostics of Mellon. And you could use something like the SAML Tracer to inspect what certificate is in use and whether it matches what you configured.

**Comment by thiema** [comment]
> 3.: If you get a signature failure I'd first investigate whether you're configured the correct assertion signing certificate from the IdP in mellon. 

Thank you for your reply ! one Additional question back:

What is correct configuration option ?
In the mod_auth_mellon Apache Config I found following Parameters:
    MellonIdPPublicKeyFile 
    MellonIdPCAFile
Are those relevant ?
And one more comment: In the meantime I went to the liblasso code and compiled a patched verision that has verification of signature of saml responses deactivated (hardcoded). This seems so "solve" my issue. However, if you can tell me a correct configuration option (either skipping of verification or providing correct public certificate), that would be great !

**Comment by thijskh** [comment]
Indeed if you would disable signature verification then it "works", however your SP is then completely vulnerable as literally anyone in the world can log into it as any user they want. So nice to check where it goes wrong, but not suited at all for production.

Yes, those options are relevant and they are documented in the README. Likely it suffices if you put the IdP's assertion signing certificate in a file you reference in `MellonIdPPublicKeyFile`.

**Comment by thiema** [comment]
The MellonIdPPublicKeyFile I already tried. Maybe I got the wrong Cert from my ADFS Provider.
You said:
> SP is then completely vulnerable as literally anyone in the world can log into it as any user they want.
Why is that the case ? 
Does the SP now accepts redirects from any IdP. Does URLs in Federation configuration xml doesn't matter ?

**Comment by thijskh** [comment]
Yes, you can check e.g. with SAML tracer which cert the IdP is using.

There's no need to accept redirects. Anyone can post an XML with their attributes to the SP if they want to. The security of the SAML websso profile depends on the XML you post being signed by the IdP.

**Comment by thiema** [comment]
Unfortunately SAML Tracer is not an option, since I cannot install browser plugins due to Windows Group Policies.

However I already configured Mellon diagnostic feature in order to see what happens behind the surface.
What I can tell is, that I already configured mellon to use the cert I got from Federation Metadata:

<KeyDescriptor use="signing"><KeyInfo xmlns="http://www.w3.org/2000/09/xmldsig#"><X509Data>...

using MellonIdPPublicKeyFile parameter.
What to conclude if that doesnt do the job ? (e.g. produce the error I have posted in the title of that issue.)
